allennlp/training/metrics.py

```diff
@@ -0,0 +1,190 @@
+        self.total_counts = 0.0
+
+    def __call__(self,
+                 predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor] = None):
+        """
+        Parameters
+        ----------
+        predictions : ``torch.Tensor``, required.
+            A tensor of predictions of shape (batch_size, ..., num_classes).
+        gold_labels : ``torch.Tensor``, required.
+            A tensor of integer class label of shape (batch_size, ...). It must be the same
+            shape as the ``predictions`` tensor without the ``num_classes`` dimension.
+        mask: ``torch.Tensor``, optional (default = None).
+            A masking tensor the same size as ``gold_labels``.
+        """
+        num_classes = predictions.size(-1)
+        if (gold_labels >= num_classes).any():
+            raise ConfigurationError("A gold label passed to F1Measure contains an id >= {}, "
+                                     "the number of classes.".format(num_classes))
+        if mask is None:
+            mask = torch.ones(gold_labels.size())
+        mask = mask.float()
+        gold_labels = gold_labels.float()
+        positive_label_mask = gold_labels.eq(self._positive_label).float()
+        negative_label_mask = 1.0 - positive_label_mask
+
+        argmax_predictions = predictions.topk(1, -1)[1].float().squeeze(-1)
+
+        # True Negatives: correct non-positive predictions.
+        correct_null_predictions = (argmax_predictions !=
+                                    self._positive_label).float() * negative_label_mask
+        self.true_negatives += (correct_null_predictions.float() * mask).sum()
+
+        # True Positives: correct positively labeled predictions.
+        correct_non_null_predictions = (argmax_predictions ==
+                                        self._positive_label).float() * positive_label_mask
+        self.true_positives += (correct_non_null_predictions * mask).sum()
+
+        # False Negatives: incorrect negatively labeled predictions.
+        incorrect_null_predictions = (argmax_predictions !=
+                                      self._positive_label).float() * positive_label_mask
+        self.false_negatives += (incorrect_null_predictions * mask).sum()
+
+        # False Positives: incorrect positively labeled predictions
+        incorrect_non_null_predictions = (argmax_predictions ==
+                                          self._positive_label).float() * negative_label_mask
+        self.false_positives += (incorrect_non_null_predictions * mask).sum()
+
+    def get_metric(self, reset: bool = False):
+        """
+        Returns
+        -------
+        Returns a tuple of the following metrics based on the accumulated count statistics:
+        precision : float
+        recall : float
+        f1-measure : float
+        """
+        recall = float(self.true_positives) / float(self.true_positives + self.false_negatives + 1e-13)
```

Comment:
I'm just being OCD here, but when I see this order I _really_ want to move the `recall` line to below the `precision` line, so that the order matches what's in the docstring and the return value.  Feel free to just laugh at me and leave it as it is, though...